lib/charms/observability_libs/v0/kubernetes_compute_resources_patch.py

```diff
@@ -0,0 +1,490 @@
+# Copyright 2022 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""# KubernetesComputeResourcesPatch Library.
+
+This library is designed to enable developers to more simply patch the Kubernetes compute resource
+limits and requests created by Juju during the deployment of a sidecar charm.
+
+When initialised, this library binds a handler to the parent charm's `config-changed` event.
+The config-changed event is used because it is guaranteed to fire on startup, on upgrade and on
+pod churn. Additionally, resource limits may be set by charm config options, which would also be
+caught out-of-the-box by this handler. The handler applies the patch to the app's StatefulSet.
+This should ensure that the resource limits are correct throughout the charm's life.Additional
+optional user-provided events for re-applying the patch are supported but discouraged.
+
+The constructor takes a reference to the parent charm, a 'limits' and a 'requests' dictionaries
+that together define the resource requirements. For information regarding the `lightkube`
+`ResourceRequirements` model, please visit the `lightkube`
+[docs](https://gtsystem.github.io/lightkube-models/1.23/models/core_v1/#resourcerequirements).
+
+
+## Getting Started
+
+To get started using the library, you just need to fetch the library using `charmcraft`. **Note
+that you also need to add `lightkube` and `lightkube-models` to your charm's `requirements.txt`.**
+
+```shell
+cd some-charm
+charmcraft fetch-lib charms.observability_libs.v0.kubernetes_compute_resources_patch
+cat << EOF >> requirements.txt
+lightkube
+lightkube-models
+EOF
+```
+
+Then, to initialise the library:
+
+```python
+# ...
+from charms.observability_libs.v0.kubernetes_compute_resources_patch import (
+    KubernetesComputeResourcesPatch
+)
+
+class SomeCharm(CharmBase):
+  def __init__(self, *args):
+    # ...
+    self.resources_patch = KubernetesComputeResourcesPatch(
+        self,
+        "container-name",
+        limits_func=lambda: {"cpu": "1", "mem": "2Gi"},
+        requests_func=lambda: {"cpu": "1", "mem": "2Gi"}
```

Comment:
Now that I am integrating this lib into Loki charm I found weird that `KubernetesComputeResourcesPatch` constructor requires a function for `limits_func` and `request_func`.

Why don't we just ask for a dictionary instead?

Reply:
It used to be a dict, and recently changed to a func.
Using "dependency injection" made more sense to me because:
- Our convention is to instantiate charm libs in the charm's constructor. The args needed for init are coming raw from what the user set as config options. Processing it in the constructor, obviating, config-changed event, is not easy.
- This way we won't need an "update_X" method we would need to call.
- Bonus: harness currently doesn't support charm reinit between events, and with this approach there are no issue.

Reply:
mmmm maybe in this case we should revisit our convention about instantiating charm libs in the constructor....

For instance, I was looking the Alertmanager [PR #87](https://github.com/canonical/alertmanager-k8s-operator/pull/87) and, from my point of view, the code is even more complex to understand:

```python
        self.resources_patch = KubernetesComputeResourcesPatch(
            self,
            self._container_name,
            limits_func=lambda: self._resource_limit_from_config(),
            requests_func=lambda: limits_to_requests_scaled(
                self._resource_limit_from_config(),
                ResourceSpecDict(cpu="0.25", memory="200Mi"),
                0.8,
            ),
        )
```

Reply:
Ideally it will be simplified to
```python
        self.resources_patch = KubernetesComputeResourcesPatch(
            self,
            self._container_name,
            limits_func=lambda: self._resource_limit_from_config(),
            requests_func=lambda: ResourceSpecDict(cpu="0.25", memory="200Mi"),
        )
```
if we decide to not scale the requests.

I am happy to discuss alternative designs.

Reply:
With the last changes this doc section should be updated.

Reply:
Fixed in next commit.